File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> list = new ArrayList();
        if(s.length()<p.length()){
            return list;
        }
        int[] pCount = new int[26];
        int[] window = new int[26];
        for(char c:p.toCharArray()){
            pCount[c-'a']++;
        }

        for(int i=0;i<p.length();i++){
            window[s.charAt(i)-'a']++;
        }

        if(Arrays.equals(window,pCount)){
            list.add(0);
        }

        for(int i=p.length();i<s.length();i++){
            window[s.charAt(i)-'a']++;

            window[s.charAt(i - p.length()) - 'a']--;

            if(Arrays.equals(window,pCount)){
                list.add(i-p.length()+1);
            }
        }
        return list;
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> freq;
        unordered_map<char, int> window;
        vector<int> ans;
        for (char c : p) {
            freq[c]++;
        }
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            window[s[right]]++;
            if (right - left + 1 > p.length()) {
                window[s[left]]--;
                if (window[s[left]] == 0) {
                    window.erase(s[left]);
                };
                left++;
            }
            if (right - left + 1 == p.length() && window == freq) {
                ans.push_back(left);
            }
        }

        return ans;
    }
};